### PR TITLE
Port `all` and `any` full reductions to structured kernels.

### PR DESCRIPTION
**Pull request by ysiraichi**
**Small BC-breaking change for `torch.all/any` interaction with uint8 and zero-dimensional tensors**

Before this PR:
```
>>> torch.all(torch.tensor(42, dtype=torch.uint8))
tensor(1, dtype=torch.uint8)
>>> torch.all(torch.tensor(42, dtype=torch.uint8), dim=0)
tensor(42, dtype=torch.uint8)
```
After this PR:
```
>>> torch.all(torch.tensor(42, dtype=torch.uint8))
tensor(1, dtype=torch.uint8)
>>> torch.all(torch.tensor(42, dtype=torch.uint8), dim=0)
tensor(1, dtype=torch.uint8)
```
This behavior is now consistent between `torch.all(x)` and `torch.all(x, dim=0)`

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#64642 Port `all` and `any` full reductions to structured kernels.**

Tracking issue: #55070 

This PR creates out overloads for both `all` and `any` kernels (full reduction overload),
and ports them to structured kernels.

Differential Revision: [D30867354](https://our.internmc.facebook.com/intern/diff/D30867354)

cc @ezyang @gchanan

### CODE REVIEW

**Comment by ysiraichi** [comment]
CI failure not related (see #64595).

**Comment by ezyang** [comment]
How come you're allowed to delete the calls to `_dimreduce_return_trivial` ?

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867354).

**Comment by ysiraichi** [comment]
> How come you're allowed to delete the calls to `_dimreduce_return_trivial` ?

As far as I understand, that function does two things:
- resize the result tensor (the meta function takes care of it)
- fill the result tensor with the appropriate value (the `_<name>` function takes care of it)

There was one case missing: for `self.numel() == 1 && self.ndimension() == 0`. Therefore, I added the following to the `_<name>` functions: `iter.numel() == 1` (I don't think we need to check for the dimension here, since we already did the resizing).

With this, I believe the implementation is equivalent.

**Comment by ysiraichi** [comment]
@ezyang @bdhirsh
Apparently, `all` and `all.dim` (same for `any`) had slightly different behavior for 0-dimensional tensors. Which should be the correct result when executing `torch.all(torch.tensor(42, dtype=torch.uint8))`?

| (a) | (b) |
|-----|-----|
| `tensor(1, dtype=torch.uint8)` | `tensor(42, dtype=torch.uint8)` |

Before this PR, `all(tensor)` returned **(a)**, while `all(tensor, dim=0)` returned **(b)**. In order to be consistent, I'm defaulting to **(a)** (feels right), but have no strong feelings about this one. What do you think?

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867354).

**Comment by ezyang** [comment]
I agree with your analysis. I'll mark this as BC-breaking just to be safe tho.

**Comment by ezyang** [comment]
@ngimel pointed out some relevant prior art https://github.com/pytorch/pytorch/pull/47878 but I think it doesn't apply here (as you are not changing the dtype)

**Comment by ngimel** [comment]
Right, and without changing dtype I agree that `1` is more consistent than non-1.
